File: packages/python/helix_sdk/services/provider.py
# packages/python/helix_sdk/services/provider.py
from __future__ import annotations

from typing import Any

from ..capabilities import CapabilityGate
from ..transport import RpcClient


class ProviderService:
    def __init__(self, rpc: RpcClient, gate: CapabilityGate) -> None:
        self._rpc = rpc
        self._gate = gate

    async def list(self) -> dict[str, Any]:
        self._gate.require("provider.list")
        return await self._rpc.call("provider.list", {})

File: packages/typescript/src/services/provider.ts
// packages/typescript/src/services/provider.ts
import type { RpcClient } from "../transport/rpc.js";
import type { CapabilityGate } from "../capabilities.js";
import type { LlmProviderView } from "./agent.js";

export interface ProviderService {
  list(): Promise<{ providers: LlmProviderView[] }>;
}

export function createProviderService(rpc: RpcClient, gate: CapabilityGate): ProviderService {
  return {
    async list() {
      gate.require("provider.list");
      return rpc.call<{ providers: LlmProviderView[] }>("provider.list", {});
    },
  };
}
